src/unitdata.rs

```diff
@@ -0,0 +1,281 @@
+    /// Returns number of rows deleted
+    pub fn unsetrange(&self,
+                      keys: Option<Vec<String>>,
+                      prefix: Option<String>)
+                      -> Result<u32, JujuError> {
+        let deleted = String::from("\"DELETED\"");
+        let revision = self.revision.clone().unwrap_or(0);
+        let prefix = prefix.unwrap_or("".to_string());
+
+        match keys {
+            Some(keys) => {
+                let mut question_marks = Vec::new();
+                for _ in 0..keys.len() {
+                    question_marks.push("?");
+                }
+                let mut values: Vec<&rusqlite::types::ToSql> = Vec::new();
+                for key in &keys {
+                    values.push(key);
+                }
+                let delete_query = format!("delete from kv where key in ({})",
+                                           question_marks.join(","));
+                let rowcount = self.conn
+                    .execute(&delete_query, &values[..])?;
+
+                if self.revision.is_some() && rowcount > 0 {
+                    let field_list = iter::repeat("(?, ?, ?)")
+                        .take(keys.len())
+                        .collect::<Vec<_>>()
+                        .join(",");
+                    //key, self.revision, String::from("\"DELETED\""
+                    let mut values: Vec<&rusqlite::types::ToSql> = Vec::new();
+                    for key in &keys {
+                        values.push(key);
+                        values.push(&revision);
+                        values.push(&deleted);
+                    }
+                    let _ = self.conn
+                        .execute(&format!("insert into kv_revisions values {}", field_list),
+                                 &values[..]);
+                }
+                Ok(rowcount as u32)
+            }
+            None => {
+                let rowcount = self.conn
+                    .execute("delete from kv where key like ?",
+                             &[&format!("{}%", prefix)])?;
+                if self.revision.is_some() && rowcount > 0 {
+                    self.conn
+                        .execute("insert into kv_revisions values (?, ?, ?)",
+                                 &[&format!("{}%", prefix), &revision, &deleted])?;
+                }
+                Ok(rowcount as u32)
+            }
+        }
+    }
+
+    /// Set a value in the database.
+    pub fn set<T>(&self, key: &str, value: T) -> Result<(), JujuError>
+        where T: Serialize
+    {
+        let serialized = serde_json::to_string(&value)?;
+
+        let exists: bool = self.conn
+            .query_row("select exists(select data from kv where key=?)",
+                       &[&key],
+                       |row| row.get(0))?;
+        match exists {
+            true => {
+                self.conn.execute("update kv set data = ? where key = ?", &[&serialized, &key])?;
```

Comment:
YOu may want to also bail up updating when there is no change: https://bazaar.launchpad.net/~charm-helpers/charm-helpers/devel/view/head:/charmhelpers/core/unitdata.py#L280